rectools/models/bert4rec.py

```diff
@@ -0,0 +1,197 @@
+        yw = np.zeros((batch_size, self.session_max_len))
+        for i, (ses, ses_weights) in enumerate(batch):
+            masked_session, target = self._mask_session(ses)
+            x[i, -len(ses) + 1 :] = masked_session[:-1]  # ses: [session_len] -> x[i]: [session_max_len]
+            y[i, -len(ses) + 1 :] = target[:-1]  # ses: [session_len] -> y[i]: [session_max_len]
+            yw[i, -len(ses) + 1 :] = ses_weights[:-1]  # ses_weights: [session_len] -> yw[i]: [session_max_len]
+
+        return torch.LongTensor(x), torch.LongTensor(y), torch.FloatTensor(yw)
+
+    def _collate_fn_recommend(self, batch: List[Tuple[List[int], List[float]]]) -> torch.LongTensor:
+        """Right truncation, left padding to session_max_len"""
+        x = np.zeros((len(batch), self.session_max_len))
+        for i, (ses, _) in enumerate(batch):
+            session = ses.copy()
+            session = session[1:] + [1]
+            x[i, -len(ses) :] = session[-self.session_max_len :]
+        return torch.LongTensor(x)
+
+
+class BERT4RecTransformerLayers(TransformerLayersBase):
+    """TODO"""
+
+    def __init__(
+        self,
+        n_blocks: int,
+        n_factors: int,
+        n_heads: int,
+        dropout_rate: float,
+    ):
+        super().__init__()
+        self.n_blocks = n_blocks
+        self.multi_head_attn = nn.ModuleList(
+            [nn.MultiheadAttention(n_factors, n_heads, dropout_rate, batch_first=True) for _ in range(n_blocks)]
+        )
+        self.layer_norm1 = nn.ModuleList([nn.LayerNorm(n_factors) for _ in range(n_blocks)])
+        self.dropout1 = nn.ModuleList([nn.Dropout(dropout_rate) for _ in range(n_blocks)])
+        self.layer_norm2 = nn.ModuleList([nn.LayerNorm(n_factors) for _ in range(n_blocks)])
+        self.feed_forward = nn.ModuleList(
+            [PointWiseFeedForward(n_factors, n_factors * 4, dropout_rate, torch.nn.GELU()) for _ in range(n_blocks)]
+        )
+        self.dropout2 = nn.ModuleList([nn.Dropout(dropout_rate) for _ in range(n_blocks)])
+        self.dropout3 = nn.ModuleList([nn.Dropout(dropout_rate) for _ in range(n_blocks)])
+
+    def forward(
+        self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: torch.Tensor, key_padding_mask: torch.Tensor
+    ) -> torch.Tensor:
+        """TODO"""
+        for i in range(self.n_blocks):
+            mha_input = self.layer_norm1[i](seqs)
+            mha_output, _ = self.multi_head_attn[i](
+                mha_input,
+                mha_input,
+                mha_input,
+                attn_mask=attn_mask,
+                key_padding_mask=key_padding_mask,
+                need_weights=False,
+            )
+            seqs = seqs + self.dropout1[i](mha_output)
+            ff_input = self.layer_norm2[i](seqs)
+            ff_output = self.feed_forward[i](ff_input)
+            seqs = seqs + self.dropout2[i](ff_output)
+            seqs = self.dropout3[i](seqs)
+
+        return seqs
+
+
```

Comment:
Let's add:
`# TODO: test with torch.nn.Linear and cross-entropy loss as in https://github.com/jaywonchung/BERT4Rec-VAE-Pytorch/blob/f66f2534ebfd937778c7174b5f9f216efdebe5de/models/bert.py#L11C1-L11C67`

Reply:
fixed